hail/src/main/scala/is/hail/utils/StackSafe.scala

```diff
@@ -0,0 +1,176 @@
+package is.hail.utils
+
+import scala.annotation.tailrec
+import scala.collection.generic.{CanBuild, CanBuildFrom}
+
+object StackSafe {
+
+  def done[A](result: A): StackFrame[A] = Done(result)
+
+  def call[A](body: => StackFrame[A]): StackFrame[A] =
+    Thunk(() => body)
+
+  object StackFrame {
+    @tailrec def run[A](frame: StackFrame[A]): A = frame match {
+      case Done(result) => result
+      case Thunk(thunk) => run(thunk())
+      case more: More[x, A] => run(more.advance())
+    }
+  }
+
+  abstract class StackFrame[A] {
+    def flatMap[B](f: A => StackFrame[B]): StackFrame[B] = this match {
+      case Done(result) => f(result)
+      case thunk: Thunk[A] =>
+        val cell = new ContCell[A, B](f)
+        new More[A, B] (thunk, cell, cell)
```

Comment:
Can you do:

```
      case thunk: Thunk[A] =>
        Thunk(() => f(run(thunk)))
```

here?  I can't quite puzzle out the motivation for More and ContCell.

Reply:
Nevermind, I see, the thunk bodies get nested.

Reply:
To minimize the number of `Thunk` closures that are allocated, we can sometimes make multiple `flatMap` calls without any `call` wrappers. E.g. in the parser, I have
```
  def ir_value_exprs(env: IRParserEnvironment)(it: TokenIterator): StackFrame[Array[IR]] = {
    punctuation(it, "(")
    for {
      irs <- ir_value_children(env)(it)
      _ = punctuation(it, ")")
    } yield irs
  }

  def ir_value_children(env: IRParserEnvironment)(it: TokenIterator): StackFrame[Array[IR]] =
    repUntil(it, ir_value_expr(env), PunctuationToken(")"))

  def ir_value_expr(env: IRParserEnvironment)(it: TokenIterator): StackFrame[IR] = {
    punctuation(it, "(")
    for {
      ir <- call(ir_value_expr_1(env)(it))
      _ = punctuation(it, ")")
    } yield ir
  }

  def ir_value_expr_1(env: IRParserEnvironment)(it: TokenIterator): StackFrame[IR] = {
    identifier(it) match {
      ...
    }
  }
```
where the only `call` is in `ir_value_expr`, because any recursive call to `ir_value_expr_1` must go through that. So we can end up with chains of flatMaps of the shape `sf.flatMap(cont1).flatMap(cont2)...`, and for efficiency I accumulate the continuations in a linked list `More(sf, cont1, contN)`.